### DISCUSSION TITLE
Colorblind accessibility discussion

### DISCUSSION THREAD

**Discussion by epicEaston197** [Yellow-Dog-Man/Resonite-Issues#1959] [General]
This discussion started from issue #1947 this discussion is a continuation of this issue

here's a quote from the issue i mentioned

from Frooxius
>a lot of the stuff proposed here aren't really settings and there's not a single actionable item from this.
Once we figure out what concrete changes and additions to make, we can make those into distinct issues.

**Comment by ColinTimBarndt** [comment]
> A super quick and dirty mockup I made of a potential improvement for both colorblind users and also just generally improving protoflux learnability
> ![Untitled](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/89536622/2c28c1ee-a87f-4013-b757-2842be12af3a)

I'd recommend using color blindness simulations instead of simple black/white filters. Those issues can be partially improved by changing the palette depending on the dichromacy of the user. Monochromacy (black and white) is very rare while [Prot-/Deuteranomaly](2) is fairly common, especially for male users. Additionally, a [second channel](3) should be used (e.g. shapes of inputs or a texture on the wire).

Second channels are one recommendation made by the [Accessible Player Experiences (APX)](1) guidelines (also quoted by [the paper I referred to](4)).

[1]: https://accessible.games/accessible-player-experiences/
[2]: https://en.m.wikipedia.org/wiki/Color_blindness
[3]: https://accessible.games/accessible-player-experiences/access-patterns/second_channel/
[4]: https://dl.acm.org/doi/10.1145/3611026

**Reply by epicEaston197** [reply]
from Frooxius
>What we were wanting to do for types that I think would be better approach is to introduce unique black & white patterns for each common type, which would be displayed there.

I think unique black and white patterns would suffice along with some color palette modifications

**Reply by ColinTimBarndt** [reply]
Quoting @Frooxius:
> The mockup only works for this particular node, because it doesn't have any outputs, but it will not work for lots of others, which already use the space you used for the type to have labels for the output connectors.

I believe that second channels would fix this issue. They're already used for higher-dimension types like int3 (the lines on the connector). Maybe a simple (procedural) texture could be generated for each type.

**Reply by ColinTimBarndt** [reply]
> I think unique black and white patterns would suffice along with some color palette modifications

I see, we were typing this concurrently 😂. I agree.

**Reply by JackTheFoxOtter** [reply]
Adding the type name to the sockets on the nodes makes sense, but it probably can't be under the label, that won't work if the node has an output socket on the opposite height.

Just putting "\<name\> (\<type\>)" on every socket label might be the easiest to implement.

**Reply by shadowpanther** [reply]
This might also overencumber the text for some compound types.

**Comment by ColinTimBarndt** [comment]
Here's my WIP designs that I came up with. These could be added to the wires as a repeating background.
![Resonite Protoflux Accessibility](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/26167910/eb50a681-a818-477f-830e-982be646b2c5)
[Inkscape SVG](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/26167910/30de0567-95c9-4ef0-bb13-3b0b2b762e1b)

From top to bottom:
- Int Design 1 (integer wave)
- Int Design 2 (binary counter)
- Float Design 1 (Bèzier wave)

I think it's not necessary to make the subtypes (float vs. double, int vs. long vs. short vs. byte) differentiable by pattern as they rarely mix and currently have the same colors in the game (when clamped to 255 by non-hdr screens).

**Reply by JackTheFoxOtter** [reply]
I'm not sure this scales very well. Currently the colors for the wires are generated, this would require there to be a texture that manually audited for every wire type, which is basically impossible since you can make pretty much any arbitrary object type for ProtoFlux.

It might be a bit boring almost, but perhaps just adding faint text to the wires denoting their type would make more sense to have as an option.

**Reply by epicEaston197** [reply]
>It might be a bit boring almost, but perhaps just adding faint text to the wires denoting their type would make more sense to have as an option.

This would mean that Resonite would need to have a generic text to image system basically turning text into a image form of it which might be useful in some cases outside of color blindness

**Reply by JackTheFoxOtter** [reply]
I mean, I don't think it makes sense to focus on implementation details here, this is more a general feature design discussion, is it not?

**Reply by ColinTimBarndt** [reply]
Resonite has the colors for those basic types hard-coded. I think it makes sense to have meaningful patterns for those common types as well. Everything else could just be generated.

**Reply by JackTheFoxOtter** [reply]
IIRC there are very few colors that are actually hardcoded. I think it makes more sense to have them consistent. Like, either unique patterns on every wire, or type name on every wire, but not mixing both.

**Reply by anquietas-sys** [reply]
I really like the pattern idea

**Reply by ColinTimBarndt** [reply]
Hardcoded colors can be seen here: https://wiki.resonite.com/Module:ProtoFlux_Type_Color
Basically every common type is hard-coded.

**Reply by Banane9** [reply]
I like the idea of representing the discrete vs. fluid nature of ints and floats as square waves and continuous waves respectively. Perhaps the bigger ("more precise") ones could have more steps. So bool could be just alternating light / dark, while bytes, ints, and longs get increasingly more "steps" in the wave. For short / float / double, it could be represented by the frequency instead 🤔

**Reply by epicEaston197** [reply]
Beyond color blindness accessibility this is just a really nice visual change overall

**Reply by JackTheFoxOtter** [reply]
> Beyond color blindness accessibility this is just a really nice visual change overall

It would have to be an option, because I wouldn't want this on for myself. I think it would mainly just make the code a lot messier to look at for me.
